src/persist-txn/src/txn_read.rs

```diff
@@ -321,19 +340,85 @@ enum WaitTs<T> {
 }
 
 #[derive(Debug)]
-struct TxnsReadTask<T: Timestamp + Lattice + Codec64, C: TxnsCodec = TxnsCodecDefault> {
+struct TxnsReadTask<T: Timestamp + Lattice + Codec64> {
     rx: mpsc::Receiver<TxnsReadCmd<T>>,
-    cache: TxnsCache<T, C>,
+    cache: TxnsCacheState<T>,
+    pending_waits: Vec<PendingWait<T>>,
+    frontier: T,
+}
+
+/// A pending "wait" notification that we will complete once the frontier
+/// advances far enough.
+#[derive(Debug)]
+struct PendingWait<T: Timestamp + Lattice + Codec64> {
+    ts: WaitTs<T>,
+    tx: Option<oneshot::Sender<()>>,
 }
 
-impl<T, C> TxnsReadTask<T, C>
+impl<T: Timestamp + Lattice + Codec64> PendingWait<T> {
+    /// Returns `true` if this [PendingWait] is completed.
+    ///
+    /// A pending wait is completed when the frontier advances far enough or
+    /// when the receiver side hangs up.
+    fn maybe_complete(&mut self, frontier: &T) -> bool {
+        if self.tx.is_none() {
+            // Already completed!
+            return true;
+        }
+
+        match &self.ts {
+            WaitTs::GreaterEqual(ts) => {
```

Comment:
Let's pull this out into a method on `WaitTs` that takes the frontier and returns a bool

Reply:
:+1: